adminSiteClient/CreateDataInsightModal.tsx

```diff
@@ -0,0 +1,935 @@
+
+type FormFieldName =
+    | "title"
+    | "authors"
+    | "grapherUrl"
+    | "narrativeChart"
+    | "figmaUrl"
+    | "imageFilename"
+    | "imageAltText"
+type ImageFormFieldName = "imageFilename" | "imageAltText"
+
+type FormData = Partial<Record<FormFieldName, string>>
+type FormDataWithTitle = RequiredBy<FormData, "title">
+type FormDataWithImageFilename = RequiredBy<FormData, "imageFilename">
+
+type DataInsightCreationResponse =
+    | { success: true; gdocId: string }
+    | { success: false; errorMessage: string }
+
+interface NarrativeChart {
+    name: string
+    configId: string
+    title: string
+}
+
+type ImageSource = "figma" | "narrativeChart" | "grapher" | "explorer"
+
+const spinnerIcon = <FontAwesomeIcon icon={faSpinner} size="sm" spin />
+const checkIcon = <FontAwesomeIcon icon={faCheck} size="sm" />
+const robotIcon = <FontAwesomeIcon icon={faRobot} size="sm" />
+const successIcon = <FontAwesomeIcon icon={faCheckCircle} size="sm" />
+const failIcon = <FontAwesomeIcon icon={faCircleXmark} size="sm" />
+const downloadIcon = <FontAwesomeIcon icon={faDownload} size="sm" />
+
+export function CreateDataInsightModal(props: {
+    // modal settings
+    description?: string
+    initialValues?: FormData
+    hiddenFields?: Omit<FormFieldName, ImageFormFieldName>[]
+
+    // data
+    narrativeChart?: NarrativeChart // create a DI for this narrative chart if given
+
+    // actions
+    closeModal: () => void
+    onFinish?: (response: DataInsightCreationResponse) => void
+}) {
+    const { admin } = useContext(AdminAppContext)
+
+    const initialValues = useMemo(
+        () => props.initialValues ?? {},
+        [props.initialValues]
+    )
+    const hiddenFields = useMemo(
+        () => props.hiddenFields ?? [],
+        [props.hiddenFields]
+    )
+
+    const [form] = Form.useForm<FormData>()
+    const [formData, setFormData] = useState<FormData>(initialValues)
+
+    // tracks progress of various tasks like uploading an image or creating a DI
+    const { progress, setProgress, updateProgress } = useProgress(
+        "createDI",
+        "uploadImage",
+        "loadFigmaImage",
+        "suggestAltText"
+    )
+
+    // loaded from Figma if a Figma URL is provided
+    const [figmaImageUrl, setFigmaImageUrl] = useState<string | undefined>()
+
+    // used for autocompletion
+    const [allNarrativeCharts, setAllNarrativeCharts] = useState<
+        NarrativeChart[]
+    >([])
+
+    const allNarrativeChartsMap = useMemo(
+        () => new Map(allNarrativeCharts.map((chart) => [chart.name, chart])),
+        [allNarrativeCharts]
+    )
+
+    const narrativeChart = formData.narrativeChart
+        ? allNarrativeChartsMap.get(formData.narrativeChart)
+        : props.narrativeChart
+
+    const getDataInsightImageUrl = ({ cache = false } = {}):
+        | { source: ImageSource; url: string }
+        | undefined => {
+        if (figmaImageUrl) return { source: "figma", url: figmaImageUrl }
```

Comment:
[me reading these 4 lines](https://www.youtube.com/watch?v=rQgXkNYHhKU&t=15s) 😹 

I think deconstructing the options param makes this a little more verbose than it needs to be, especially because it defaults to false in `makePngUrlForNarrativeChart` anyway.

In fact, can cache _ever_ be set to true?

Reply:
In almost all cases we want an uncached version. I added the cache option later because the alt text suggestion sometimes times out for our Grapher thumbnails – using the cached version helps

Reply:
I don't expect it to be unreliable on the images index page because there, we serve from Cloudflare, which is presumably much faster than our workers (especially if we don't rely on the cache, as then the workers need to render Grapher first)